### PR TITLE
Use transparent functors in placeholder expressions.

### PR DESCRIPTION
**Pull request by allisonvacanti**
Fixes and adds regression tests for #1178 & #1229.

### CODE REVIEW

**Comment by mfrancis95** [comment]
Is the plan to bring #1186 back after this is merged in?

**Comment by allisonvacanti** [comment]
That commit was only reverted on a staging branch for an upcoming release, which is only taking critical bugfixes right now.

#1186 is still in `main` and this patch fixes the related regressions by making the placeholder implementation more robust.

**Comment by allisonvacanti** [comment]
Rebased on master, started DVS CL 28862936.

**Comment by allisonvacanti** [comment]
Confirmed that this change doesn't significantly impact compile times.

Restarted DVS on same CL after refactoring the new static assert in `actor`.

**Comment by allisonvacanti** [comment]
@dkolsen-pgi can you review this when you get a chance? It addresses your issue with `thrust::find`, #1229.

File: thrust/detail/functional/operators/operator_adaptors.h [hunk 1]
```diff
@@ -27,87 +32,104 @@ namespace detail
...
-// this thing (which models Eval) is an adaptor for the binary
-// functors inside functional.h
-template<template<typename> class BinaryOperator>
-  struct binary_operator
+  result_type<Env> eval(Env&& e) const
+  THRUST_RETURNS(UnaryFunctor{}(thrust::get<0>(THRUST_FWD(e))))
+};
+
+
+// Adapts a transparent unary functor from functional.h (e.g. thrust::less<>)
```
Comment (dkolsen-pgi) [inline thrust/detail/functional/operators/operator_adaptors.h:80]: "unary" should be "binary" in the comment.

Comment (allisonvacanti) [reply thrust/detail/functional/operators/operator_adaptors.h]: Fixed.

**Comment by allisonvacanti** [comment]
rebased and addressed review feedback.

**Pull request merged**